// ngx-mat-timepicker/esm2022/lib/pipes/ngx-mat-timepicker-active-minute.pipe.mjs
import { Pipe } from "@angular/core";
import { NgxMatTimepickerUtils } from "../utils/ngx-mat-timepicker.utils";
import * as i0 from "@angular/core";
export class NgxMatTimepickerActiveMinutePipe {
    transform(minute, currentMinute, gap, isClockFaceDisabled) {
        if (minute == null || isClockFaceDisabled) {
            return false;
        }
        return ((currentMinute === minute) && (minute % (gap || NgxMatTimepickerUtils.DEFAULT_MINUTES_GAP) === 0));
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.5", ngImport: i0, type: NgxMatTimepickerActiveMinutePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe }); }
    static { this.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "18.0.5", ngImport: i0, type: NgxMatTimepickerActiveMinutePipe, isStandalone: true, name: "activeMinute" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.5", ngImport: i0, type: NgxMatTimepickerActiveMinutePipe, decorators: [{
            type: Pipe,
            args: [{
                    name: "activeMinute",
                    standalone: true
                }]
        }] });
