// ngx-mat-timepicker/esm2022/lib/services/ngx-mat-timepicker-locale.service.mjs
import { Inject, Injectable } from "@angular/core";
//
import { NGX_MAT_TIMEPICKER_LOCALE } from "../tokens/ngx-mat-timepicker-time-locale.token";
import * as i0 from "@angular/core";
export class NgxMatTimepickerLocaleService {
    get locale() {
        return this._locale;
    }
    constructor(initialLocale) {
        this._locale = initialLocale;
    }
    updateLocale(newValue) {
        this._locale = newValue || this._initialLocale;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.5", ngImport: i0, type: NgxMatTimepickerLocaleService, deps: [{ token: NGX_MAT_TIMEPICKER_LOCALE }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.0.5", ngImport: i0, type: NgxMatTimepickerLocaleService, providedIn: "root" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.5", ngImport: i0, type: NgxMatTimepickerLocaleService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: "root"
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [NGX_MAT_TIMEPICKER_LOCALE]
                }] }] });
