// @ngrx/eslint-plugin/src/rules/effects/no-dispatch-in-effects.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.noDispatchInEffectsSuggest = exports.noDispatchInEffects = void 0;
var path = __importStar(require("path"));
var rule_creator_1 = require("../../rule-creator");
var utils_1 = require("../../utils");
exports.noDispatchInEffects = 'noDispatchInEffects';
exports.noDispatchInEffectsSuggest = 'noDispatchInEffectsSuggest';
exports.default = (0, rule_creator_1.createRule)({
    name: path.parse(__filename).name,
    meta: {
        type: 'suggestion',
        ngrxModule: 'effects',
        hasSuggestions: true,
        docs: {
            description: '`Effect` should not call `store.dispatch`.',
            recommended: 'warn',
            suggestion: true,
        },
        schema: [],
        messages: (_a = {},
            _a[exports.noDispatchInEffects] = 'Calling `store.dispatch` in `Effect` is forbidden.',
            _a[exports.noDispatchInEffectsSuggest] = 'Remove `store.dispatch`.',
            _a),
    },
    defaultOptions: [],
    create: function (context) {
        var _a;
        var _b = (0, utils_1.getNgRxStores)(context).identifiers, identifiers = _b === void 0 ? [] : _b;
        var storeNames = identifiers.length > 0 ? (0, utils_1.asPattern)(identifiers) : null;
        if (!storeNames) {
            return {};
        }
        return _a = {},
            _a[(0, utils_1.dispatchInEffects)(storeNames)] = function (node) {
                var nodeToReport = getNodeToReport(node);
                context.report({
                    node: nodeToReport,
                    messageId: exports.noDispatchInEffects,
                    suggest: [
                        {
                            messageId: exports.noDispatchInEffectsSuggest,
                            fix: function (fixer) { return fixer.remove(nodeToReport); },
                        },
                    ],
                });
            },
            _a;
    },
});
function getNodeToReport(node) {
    var parent = node.parent;
    var grandParent = parent.parent;
    return grandParent &&
        ((0, utils_1.isArrowFunctionExpression)(grandParent) || (0, utils_1.isReturnStatement)(grandParent))
        ? node
        : parent;
}
