// @ngrx/eslint-plugin/src/rules/store/no-multiple-global-stores.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.noMultipleGlobalStoresSuggest = exports.noMultipleGlobalStores = void 0;
var path = __importStar(require("path"));
var rule_creator_1 = require("../../rule-creator");
var utils_1 = require("../../utils");
exports.noMultipleGlobalStores = 'noMultipleGlobalStores';
exports.noMultipleGlobalStoresSuggest = 'noMultipleGlobalStoresSuggest';
exports.default = (0, rule_creator_1.createRule)({
    name: path.parse(__filename).name,
    meta: {
        type: 'suggestion',
        hasSuggestions: true,
        ngrxModule: 'store',
        docs: {
            description: 'There should only be one global store injected.',
            recommended: 'warn',
            suggestion: true,
        },
        schema: [],
        messages: (_a = {},
            _a[exports.noMultipleGlobalStores] = 'Global store should be injected only once.',
            _a[exports.noMultipleGlobalStoresSuggest] = 'Remove this reference.',
            _a),
    },
    defaultOptions: [],
    create: function (context) {
        return {
            Program: function () {
                var e_1, _a, e_2, _b;
                var _c = (0, utils_1.getNgRxStores)(context), _d = _c.identifiers, identifiers = _d === void 0 ? [] : _d, sourceCode = _c.sourceCode;
                var flattenedIdentifiers = groupBy(identifiers).values();
                try {
                    for (var flattenedIdentifiers_1 = __values(flattenedIdentifiers), flattenedIdentifiers_1_1 = flattenedIdentifiers_1.next(); !flattenedIdentifiers_1_1.done; flattenedIdentifiers_1_1 = flattenedIdentifiers_1.next()) {
                        var identifiers_2 = flattenedIdentifiers_1_1.value;
                        if (identifiers_2.length <= 1) {
                            continue;
                        }
                        var _loop_1 = function (node) {
                            var nodeToReport = getNodeToReport(node);
                            context.report({
                                node: nodeToReport,
                                messageId: exports.noMultipleGlobalStores,
                                suggest: [
                                    {
                                        messageId: exports.noMultipleGlobalStoresSuggest,
                                        fix: function (fixer) { return getFixes(sourceCode, fixer, nodeToReport); },
                                    },
                                ],
                            });
                        };
                        try {
                            for (var identifiers_1 = (e_2 = void 0, __values(identifiers_2)), identifiers_1_1 = identifiers_1.next(); !identifiers_1_1.done; identifiers_1_1 = identifiers_1.next()) {
                                var node = identifiers_1_1.value;
                                _loop_1(node);
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (identifiers_1_1 && !identifiers_1_1.done && (_b = identifiers_1.return)) _b.call(identifiers_1);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (flattenedIdentifiers_1_1 && !flattenedIdentifiers_1_1.done && (_a = flattenedIdentifiers_1.return)) _a.call(flattenedIdentifiers_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            },
        };
    },
});
function getNodeToReport(node) {
    return node.parent && (0, utils_1.isTSParameterProperty)(node.parent) ? node.parent : node;
}
function getFixes(sourceCode, fixer, node) {
    var parent = node.parent;
    var nodeToRemove = parent && (0, utils_1.isTSParameterProperty)(parent) ? parent : node;
    return (0, utils_1.getNodeToCommaRemoveFix)(sourceCode, fixer, nodeToRemove);
}
function groupBy(identifiers) {
    return identifiers.reduce(function (accumulator, identifier) {
        var parent = (0, utils_1.isTSParameterProperty)(identifier.parent)
            ? identifier.parent.parent
            : identifier.parent;
        var collectedIdentifiers = accumulator.get(parent);
        return accumulator.set(parent, __spreadArray(__spreadArray([], __read((collectedIdentifiers !== null && collectedIdentifiers !== void 0 ? collectedIdentifiers : [])), false), [
            identifier,
        ], false));
    }, new Map());
}
