// @ngrx/eslint-plugin/src/rules/store/prefer-inline-action-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.preferInlineActionPropsSuggest = exports.preferInlineActionProps = void 0;
var path = __importStar(require("path"));
var rule_creator_1 = require("../../rule-creator");
var utils_1 = require("../../utils");
exports.preferInlineActionProps = 'preferInlineActionProps';
exports.preferInlineActionPropsSuggest = 'preferInlineActionPropsSuggest';
exports.default = (0, rule_creator_1.createRule)({
    name: path.parse(__filename).name,
    meta: {
        type: 'suggestion',
        hasSuggestions: true,
        ngrxModule: 'store',
        docs: {
            description: 'Prefer using inline types instead of interfaces, types or classes.',
            recommended: 'warn',
            suggestion: true,
        },
        schema: [],
        messages: (_a = {},
            _a[exports.preferInlineActionProps] = 'Use inline types instead of interfaces, types or classes.',
            _a[exports.preferInlineActionPropsSuggest] = 'Change to inline types.',
            _a),
    },
    defaultOptions: [],
    create: function (context) {
        var _a;
        return _a = {},
            _a[utils_1.actionCreatorPropsComputed] = function (node) {
                context.report({
                    node: node,
                    messageId: exports.preferInlineActionProps,
                    suggest: [
                        {
                            messageId: exports.preferInlineActionPropsSuggest,
                            fix: function (fixer) { return [
                                fixer.insertTextBefore(node, '{name: '),
                                fixer.insertTextAfter(node, '}'),
                            ]; },
                        },
                    ],
                });
            },
            _a;
    },
});
